### ISSUE TITLE
Touchpad module oscillation

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/firmware#364]
The touchpad module oscillates in Agent.

![2021-04-29 07 55 35](https://user-images.githubusercontent.com/300402/116807768-3f62fb80-ab35-11eb-9cc1-212116a4956c.gif)

I noticed this issue a while ago. The right keyboard half keeps querying the touchpad over I2C, and the touchpad NAKed from time to time which caused the above phenomenon in Agent. I set MANUAL_CONTROL to 1 and MODE_SELECT to Active mode in bb9bb29 which resolved the issue.

Afterward, I had a chat with Azoteq to improve the tracking accuracy of the touchpad and changed some values in a2af6d994015cc3a2dd4f9b91ca1e004ea7df95a which apparently reintroduced the NAKs. The following values were changed in the Trackpad Hardware Setup values of the Prox Hardware Settings section:

1. Pass Length: 101 -> 010
2. UpLength: 101 -> 010
3. VTrip: 0011 -> 1000
4. Opamp Bias: 2.5 uA -> 5.0 uA

According to my findings, changing either 1. or 2. negates the effect of the `MANUAL_CONTROL=1` and `MODE_SELECT=Active mode` settings.

We have to figure out which settings to use to maximize tracking accuracy and get rid of the NAKs.

**Comment by dinamix1** [comment]
Having the same issue. An added note is that that the message under the firmware tab "Touchpad is detected..." is also oscillating. Thanks!

**Comment by ghost** [comment]
I have the same issue, both on Mac OS Catalina and Arch Linux.

**Comment by mhantsch** [comment]
I had that too, but I thought it may be an electrical contact problem between module and keyboard, so I rubbed the contacts a little and it seems the problem went away.

**Comment by ghost** [comment]
> I rubbed the contacts a little and it seems the problem went away.

I tried the same, used a microfiber cloth, but the problem did not go away.

If dirty contacts were the problem, I'd expect the device to function intermittently as well, but that's not the case. I'd also expect the pulsating in the UI to be less regular.

**Comment by mhantsch** [comment]
@misterwrong Well, I'm still observant to see under which conditions the issue may reoccur.

**Comment by ehdoyle** [comment]
I'm seeing the same symptoms - Ubuntu 20.04. and the 1.5.14 Linux version of agent, with updated firmware from that agent release.
Although I have to commend support for being really quick in responding to my email and directing me here. I figured it would be a day, rather than an hour before I heard anything back.

**Comment by YokaiRick** [comment]
Got my touchpad today too. It works fine, but within the agent i have the same problem on archlinux too
I use the AppImage (tried it on two different PCs)

**Comment by brokenflea** [comment]
same issue with the touchbad i received yesterday, MacOS Big Sur (11.4) running on 1.5.14 Agent on UHKv1

**Comment by dante-12** [comment]
same issue with both the trackball & touchpad module. The modules are almost non-functional. Agent 1.5.14, Firmware 8.10.9. Windows 10.

**Comment by mondalaci** [comment]
@dante-12 What do you mean exactly by saying that "the modules are almost non-functional"? The touchpad is functional, but its tap action can't be changed from left click until this gets fixed. We've received much positive feedback regarding the trackball.

**Comment by dante-12** [comment]
@mondalaci It hints me to try again since I have two UHKs (Windows & Mac). Trackball works perfectly on Mac but Windows. On Windows, I was unable to make the mouse cursor move mostly (can move very rarely). Then, I did clean a contact metal parts on UHK for Windows, then it looks working fine last a few minutes. The oscillation issue is also gone. Thanks for your response.

**Comment by brokenflea** [comment]
just checking in, is there a fix for the oscillation on touchpad or an eta to when it'll become available ?

**Comment by mondalaci** [comment]
There's no ETA, but I'll prioritize such issues as things settle up, which may take some months.

**Comment by mattgodbolt** [comment]
> There's no ETA, but I'll prioritize such issues as things settle up, which may take some months.

Thanks for the update. The touchpad is an expensive curiousity until we can configure it, and right now we can't configure it as it keeps appearing and disappearing in the UI... so a timely fix would be lovely :) I appreciate there's lots to do though!

**Comment by Slabity** [comment]
I'm glad I'm not the only one having this issue.

If the underlying firmware issue will take a while to fix, would it be possible to allow the user to configure their modules in the Agent software without them being connected to the keyboard? That would allow us to configure the touchpad module (or any other module) without dealing with this particular issue.

**Comment by mondalaci** [comment]
@Slabity You can export your configuration, edit the relevant mapping, and import your configuration as a workaround.

**Comment by YokaiRick** [comment]
Oh, good to know. I'll try it. Thanks ^^

**Comment by cvkline** [comment]
It seems like this may take some time to fix in the Agent.

Since the export/import configuration file is just a JSON, would it be possible to publish what the trackpad configuration data structure would look like (keywords, values, etc) so that the more brave among us could at least tailor the operation of the trackpad by editing that JSON by hand?

**Comment by kareltucek** [comment]
> It seems like this may take some time to fix in the Agent.

@mondalaci requiring a module to be physically present seems pretty stupid to me. As a user, I would expect to be able to edit configuration of any module just via clicking on screen without having to swap modules forth and back...

(Module configuration is not stored in modules. All configuration lives in right half, so there is no reason for such artificial constraints.)

> Since the export/import configuration file is just a JSON, would it be possible to publish what the trackpad configuration data structure would look like (keywords, values, etc) so that the more brave among us could at least tailor the operation of the trackpad by editing that JSON by hand?

**edit: actually, not possible at the moment.**

Just like any other module - just a list of actions, no special settings there.

Ids seem to be:

0) right half
1) left half
2) key cluster
3) trackball
4) trackpoint
5) touchpad

Relevant part for default module settings would look like:

```
            {
              "id": 2,
              "keyActions": [
                {
                  "keyActionType": "keystroke",
                  "type": "basic",
                  "scancode": 76
                },
                {
                  "keyActionType": "keystroke",
                  "type": "basic",
                  "scancode": 42
                },
                {
                  "keyActionType": "keystroke",
                  "type": "basic",
                  "scancode": 40
                },
                {
                  "keyActionType": "mouse",
                  "mouseAction": "leftClick"
                },
                {
                  "keyActionType": "mouse",
                  "mouseAction": "middleClick"
                },
                {
                  "keyActionType": "mouse",
                  "mouseAction": "rightClick"
                }
              ]
            },
            {
              "id": 3,
              "keyActions": [
                {
                  "keyActionType": "mouse",
                  "mouseAction": "leftClick"
                },
                {
                  "keyActionType": "mouse",
                  "mouseAction": "rightClick"
                }
              ]
            },
            {
              "id": 4,
              "keyActions": [
                {
                  "keyActionType": "mouse",
                  "mouseAction": "leftClick"
                },
                {
                  "keyActionType": "mouse",
                  "mouseAction": "rightClick"
                }
              ]
            },
            {
              "id": 5,
              "keyActions": [
                {
                  "keyActionType": "mouse",
                  "mouseAction": "leftClick"
                }
              ]
            }
```

If some is missing, it apparently has to be added into all layers of a keymap in order to take effect.

**Comment by kareltucek** [comment]
But looking into the code, touchpad actions seem to be hardcoded, so even though above config shows `rightClick` in Agent, touchpad behaves still the same way.

**Comment by mondalaci** [comment]
@kareltucek The workaround I suggested is useless until #349 gets fixed, indeed.

Although it's unnecessary to have a module attached to edit its configuration, the current user interface is cleaner than any alternative I can think of, which allows the edition of any module regardless of whether it's attached. We've shipped thousands of modules, and nobody has complained about this limitation so far. Even if users knew that the possibility is technically there, I don't think it'd make a difference.

**Comment by kareltucek** [comment]
Hmm, didn't think about user interface. If I was to put it somewhere, I would probably go with a cogwheel centered right below layer bar. On click, it would open a popover that would have clickable options, one per line (reading `auto` `keycluster` `trackball` `trackpoint` `touchpad`) that could be toggled by a click. Clicking `auto` would disable everything else and similarly clicking any module would disable `auto`. Clicking any right module would disable any other right module. 

But if no-one has complained, it's indeed not worth the trouble.

**Comment by mdkcore0** [comment]
just got my modules (keycluster and the touchpad), and have the same issue on the touchpad module, tested on my desktop (Slackware 14.2) and notebook (Slackware current)

the keycluster is working as expected :)

EDIT: updating to firmware 8.10.11 seems to solve the issue for me (was on 8.10.10)

**Comment by matacino** [comment]
`had the same "feature" - an update to current agent-version 1.5.16 and firmware (8.10.12) and pulling the touchpad of, rubbing the contacts and putting it back seems to fix it.

But as other stated it would be pretty if the announced functionality

> support two-finger tapping, two-finger scrolling, and pinch to zoom. 

and

> Agent will provide module speed settings and various module-related settings. 

would come into this part of the multiverse ;-) this would be the day when it`s actually a substitute for one of this Xerox-Rats.
 
thanx @mondalaci  for his contribution to the UI

**Comment by LukePammant** [comment]
Just updated to firmware 8.10.12 and agent 1.5.16 and still getting the same issue with the touchpad module.

**Comment by mondalaci** [comment]
According to the changelog of the new firmware, it doesn't fix this issue yet. Haven't forgotten about this one, folks.

**Comment by LukePammant** [comment]
Ok thanks - I saw the two previous comments stating that it's fixed for them and got jealous. I'll standby :)

**Comment by SaswatPadhi** [comment]
I have 8.10.12 firmware and looks like this problem still exists.

EDIT: To my surprise, if I also attach the key cluster module (on the left half), this issue goes away!

**Comment by brokenflea** [comment]
issue still exists, my module is a dead paperweight since I bought it.

**Comment by mondalaci** [comment]
@brokenflea The oscillation in Agent is annoying, but your module should be useable regardless. What makes your touchpad module useless?

**Comment by brokenflea** [comment]
> @brokenflea The oscillation in Agent is annoying, but your module should be useable regardless. What makes your touchpad module useless?

Since the module oscillates there’s no way I’ve been able to get the agent to do something as simple as tap the touchpad for left click or right click or double click. The only thing I’m able to do with the module is move the cursor around which defeats the purpose of having the module in the first place.

**Comment by famuvie** [comment]
For what is worth, the **connection of a second left-side module** stops the oscillation issue and allows editing the functionality of the module.
If you have one, it can be used as a workaround in the meanwhile.

![uhk_modules](https://user-images.githubusercontent.com/5693966/152973046-919dfc1b-7c0a-4175-95cb-4ca8a1d5e62e.gif)

**Comment by kareltucek** [comment]
Does it? As far as I know, left click is still hard-coded in latest stock firmware release, so you cannot remap it anyways and there is nothing to be gained by the workaround at the moment *with stock firmware*.

(Feel free to try https://github.com/kareltucek/firmware - it contains the new touchpad code which is now heading into stock too. Left click can be remapped via Agent, UserConfig.json or via a macro command, right click can be remapped only via a macro command. Navigation modes can be reassigned or remapped by macro commands. Scroll and pinch zoom gestures are working and are connected to scroll and zoom navigation modes.)

**Comment by famuvie** [comment]
You are right. I had not actually try to remap anything. I was referring to the oscillation issue.

**Comment by bhaeussermann** [comment]
@famuvie By default - without any remapping / configuration - and the way I've been using my touchpad is:

- swiping moves the cursor,
- tap emits a left click and
- MOD + swiping scrolls.

I often use it in combination with the Mouse layer for eg. right-click or click-and-drag actions.

**Comment by mondalaci** [comment]
@kareltucek Can you still reproduce this issue? With the git master firmware and Agent, I can't reproduce it anymore.

**Comment by kareltucek** [comment]
Yes, but only when left module or left half is disconnected. 🤨

**Comment by mondalaci** [comment]
@kareltucek Interesting finding!

I can only trigger the bug when disconnecting the left half. Then I can only untrigger the bug by reconnecting the left half and the key cluster module.

Can you dig deeper? I think we're onto something.

**Comment by kareltucek** [comment]
Sure!

**Comment by mondalaci** [comment]
Resolved by #534. The newly released 9.1.1 firmware version contains the fix.
